pkg/filter/circuitbreaker/circuitbreaker.go

```diff
@@ -108,6 +108,40 @@ URLLoop:
 	return nil
 }
 
+func (spec Spec) validatePoliciesSpec() error {
+	for _, p := range spec.Policies {
+		//group of failure
+		if p.FailureRateThreshold == 0 && len(p.FailureStatusCodes) != 0 {
+			return fmt.Errorf("policy '%s' has setted failure status code, but not set failure threshold ", p.Name)
+		}
+		if p.FailureRateThreshold != 0 && len(p.FailureStatusCodes) == 0 {
+			return fmt.Errorf("policy '%s' has setted failure threshold, but not set failure status code ", p.Name)
+		}
+
+		//group of slow
+		if p.SlowCallRateThreshold != 0 && (p.SlowCallDurationThreshold == "" || strings.TrimSpace(p.SlowCallDurationThreshold) == "") {
+			return fmt.Errorf("policy '%s' has setted slow call threshold, but not set slow call duration ", p.Name)
+		}
+		if p.SlowCallRateThreshold == 0 && (p.SlowCallDurationThreshold != "" || strings.TrimSpace(p.SlowCallDurationThreshold) != "") {
+			return fmt.Errorf("policy '%s' has setted slow call duration, but not set slow call threshold ", p.Name)
+		}
+	}
+	return nil
+}
```

Comment:
Please check the `buildPolicy` function below, if `FailureRateThreshold`, `SlowCallRateThreshold`, `SlowCallDurationThreshold` is 0(or emtpy), default values will be used, so it is not an error.

And in the `handle` function, we can see `FailureRateThreshold` could be used with an empty `FailureStatusCodes`.

So I think these validations are not necessary, and we only need to update the `jsonSchema` to remove `minimum`.

Reply:
`And in the handle function, we can see FailureRateThreshold could be used with an empty FailureStatusCodes.`
Does it mean CountingNetworkError? I really missed it here. It seems to be related to the Failure policy. When the user does not specify the FailureStatusCode, the error code defined by eg is used (500/499/...). I think it would be
`if p.FailureRateThreshold != 0 && (len(p.FailureStatusCodes) == 0 || !p.CountingNetworkError)`

My understanding is that when the user configures the circuit breaker, his expectation is to trigger the circuit breaker under certain conditions, so the legality check item should be to ensure that his configuration can correctly enable the circuit breaker, instead of being silent because of the wrong configuration . As for the default value in `handle`, I wonder if the user knows and accepts the default value. It seems that the document does not mention the default value, so when the user mistakenly omits some parameters, the effect of the default value may not be what the user expects, and he may spend part of his time analyzing the positioning.
And the variable initialization in `handle` I prefer to understand it as a solution to the null pointer or to distinguish between the user's explicit setting and the eg's own setting.

ex: When users only set the SlowCallRateThreshold without setting the SlowCallDurationThreshold, eg will set the SlowCallDurationThreshold to 1 minute, and what the user sees is that the requests he tested are assumed to be returned in 5s, then the circuit breaker will not be triggered, and the same is true for Failure.

Reply:
> `And in the handle function, we can see FailureRateThreshold could be used with an empty FailureStatusCodes.` Does it mean CountingNetworkError? I really missed it here. It seems to be related to the Failure policy. When the user does not specify the FailureStatusCode, the error code defined by eg is used (500/499/...). I think it would be `if p.FailureRateThreshold != 0 && (len(p.FailureStatusCodes) == 0 || !p.CountingNetworkError)`

Yes, the check should be  `if p.FailureRateThreshold != 0 && (len(p.FailureStatusCodes) == 0 || !p.CountingNetworkError)`

> My understanding is that when the user configures the circuit breaker, his expectation is to trigger the circuit breaker under certain conditions, so the legality check item should be to ensure that his configuration can correctly enable the circuit breaker, instead of being silent because of the wrong configuration . As for the default value in `handle`, I wonder if the user knows and accepts the default value. It seems that the document does not mention the default value, so when the user mistakenly omits some parameters, the effect of the default value may not be what the user expects, and he may spend part of his time analyzing the positioning. And the variable initialization in `handle` I prefer to understand it as a solution to the null pointer or to distinguish between the user's explicit setting and the eg's own setting.
> 
> ex: When users only set the SlowCallRateThreshold without setting the SlowCallDurationThreshold, eg will set the SlowCallDurationThreshold to 1 minute, and what the user sees is that the requests he tested are assumed to be returned in 5s, then the circuit breaker will not be triggered, and the same is true for Failure.

This is a document issue, `SlowCallDurationThreshold` should be listed in [this table](https://github.com/megaease/easegress/blob/main/doc/reference/filters.md#circuitbreakerpolicy)

Reply:
@localvar @samutamm 
modified according to the discussion comments, please review again